File: usage_example/cpp/step4/src/main.cpp
#include <stdio.h>
#include "c3ga.h"

using namespace c3ga;

/*
This small example creates a line and a plane from points.
It then computes the intersection point of the line and the plane.

In this step, all types and functions are specialized.
*/
int main(int argc, char *argv[]) {

	// get five points
	normalizedPoint linePt1 = cgaPoint(1.0f, 0.0f, 0.0f);
	normalizedPoint linePt2 = cgaPoint(1.0f, 1.0f, 0.0f);
	normalizedPoint planePt1 = cgaPoint(1.0f, 2.0f, 0.0f);
	normalizedPoint planePt2 = cgaPoint(1.0f, 2.0f, 1.0f);
	normalizedPoint planePt3 = cgaPoint(0.0f, 2.0f, 1.0f);

	// output text the can be copy-pasted into GAViewer
	// warning: c_str() is not multi-threading safe (for that, use toString() instead).
	printf("linePt1 = %s,\n", linePt1.c_str()); 
	printf("linePt2 = %s,\n", linePt2.c_str());
	printf("planePt1 = %s,\n", planePt1.c_str());
	printf("planePt2 = %s,\n", planePt2.c_str());
	printf("planePt3 = %s,\n", planePt3.c_str());

	// create line and plane out of points
	line L = linePt1 ^ linePt2 ^ ni;
	plane P = planePt1 ^ planePt2 ^ planePt3 ^ ni;

	// output text the can be copy-pasted into GAViewer
	printf("L = %s,\n", L.c_str());
	printf("P = %s,\n", P.c_str());

	// compute intersection of line and plane
	flatPoint intersection = dual(L) << P;

	printf("intersection = %s,\n", intersection.c_str());

	return 0;
}

File: usage_example/java/Step4/src/Main.java
import C3ga_pkg.C3ga;
import C3ga_pkg.FlatPoint;
import C3ga_pkg.Line;
import C3ga_pkg.NormalizedPoint;
import C3ga_pkg.Plane;


public class Main {

    /*
    This small example creates a line and a plane from points.
    It then computes the intersection point of the line and the plane.

    In this step, the 'reportUsage' functionality is used to extract what
    specialized functions and types are missing.
    */
	public static void main(String[] args) {
		
        // get five points
        NormalizedPoint linePt1 = C3ga.cgaPoint(1.0f, 0.0f, 0.0f);
        NormalizedPoint linePt2 = C3ga.cgaPoint(1.0f, 1.0f, 0.0f);
        NormalizedPoint planePt1 = C3ga.cgaPoint(1.0f, 2.0f, 0.0f);
        NormalizedPoint planePt2 = C3ga.cgaPoint(1.0f, 2.0f, 1.0f);
        NormalizedPoint planePt3 = C3ga.cgaPoint(0.0f, 2.0f, 1.0f);

        // output text the can be copy-pasted into GAViewer
        System.out.println("linePt1 = " + linePt1 + ",");
        System.out.println("linePt2 = " + linePt2 + ",");
        System.out.println("planePt1 = " + planePt1 + ",");
        System.out.println("planePt2 = " + planePt2 + ",");
        System.out.println("planePt3 = " + planePt3 + ",");

        // create line and plane out of points
        Line L = C3ga.op(linePt1, C3ga.op(linePt2, C3ga.ni));

        Plane P = C3ga.op(planePt1, C3ga.op(planePt2, C3ga.op(planePt3, C3ga.ni)));

        // output text the can be copy-pasted into GAViewer
        System.out.println("L = " + L + ",");
        System.out.println("P = " + P + ",");

        // compute intersection of line and plane
        FlatPoint intersection = C3ga.lc(C3ga.dual(L), P);
    	NormalizedPoint intersectionPt = C3ga.cgaPoint(intersection); // example of how to convert to regular CGA point

        System.out.println("intersection = " + intersection + ",");
	}

}
